### PR TITLE
Revpi

### PR DESCRIPTION
**Pull request by jarongrigat**
Revpi now read out via channel name (as defined via Pictory) instead of offset number to better support readout of all the different kind of modules.
Muxer readout should work as well, needs testing.
Also database reading commands need to be adjusted for this to work...

### CODE REVIEW

File: revpi.py [hunk 1]
```diff
@@ -1,89 +1,141 @@
...
+        """
+        Set the value of a variable (most likely an output)
+        :param name: name of the variable as defined in the Pictory
+        :param value: value to be set
+        """
+        if name not in self.positions:
+            self.positions[name] = self.get_position(name)
+        offset = struct.unpack_from('>H', self.positions[name], 32)[0]
+        length = struct.unpack_from('>H', self.positions[name], 35)[0]
+        prm = (b'K'[0] << 8) + 16
```
Comment (darrylmasson) [inline revpi.py:50]: Different offset here as compared to L36 - intentional or typo?

File: revpi.py [hunk 2]
```diff
@@ -1,89 +1,141 @@
...
+        """
+        Read value of a variable
+        :param name: name of the variable as defined in the Pictory
+        """
+        if name not in self.positions:
+            self.positions[name] = self.get_position(name)
+        value = bytearray([0, 0, 0, 0])
+        offset = struct.unpack_from('>H', self.positions[name], 32)[0]
+        length = struct.unpack_from('>H', self.positions[name], 35)[0]
+        prm = (b'K'[0] << 8) + 15
```
Comment (darrylmasson) [inline revpi.py:75]: Also here

File: revpi.py [hunk 3]
```diff
@@ -1,89 +1,141 @@
...
+            struct.pack_into('>H', value, 0, offset)
+            struct.pack_into('B', value, 2, bit)
+            fcntl.ioctl(self.f, prm, value)
+            ret = value[3]
+        else:  # two bytes
+            self.f.seek(int(offset >> 8))
+            ret = int.from_bytes(self.f.read(2), 'little')
+        return ret
+
     def execute_command(self, command):
```
Comment (darrylmasson) [inline revpi.py]: The signature of this should change: `def execute_command(self, target, value)` and everything through L93 can get deleted

File: revpi.py [hunk 4]
```diff
@@ -1,89 +1,141 @@
...
+            if int(msg[2]) > (1 << 16):
+                pass
+            self.write(msg[1], msg[2])
+        else:
+            self.logger.error(f"Message starts with invalid character {msg[0]}. Allowed characters: 'r' or 'w'")
             ret['retcode'] = -1
-            return ret
+        return ret
 
     def process_one_reading(self, name, data):
```
Comment (darrylmasson) [inline revpi.py]: We also changed the name of this function to `process_one_value`

File: revpi.py [hunk 5]
```diff
@@ -1,89 +1,141 @@
...
-        if target == 'fast_cooling_valve':
+        if target in self.targets:
             # "fast_cooling_valve open" or something
-            module = 3
-            ch = 13
-            value = 1 if value == 'open' else 0
-        return self.commands['writeOutput'].format(module=module, ch=output, value=value)
+            target = self.targets[target]
+            if value in self.keywords:
+                value = self.keywords[value]
```
Comment (darrylmasson) [inline revpi.py]: We could compact these lines with something like `target=self.targets.get(target, target)` and `value=self.keywords.get(value, value)`

**Review by darrylmasson — CHANGES_REQUESTED** [review]
A couple of very minor changes necessary, otherwise it looks good.

File: revpi.py (hunk 1 above)
Comment (jarongrigat) [reply revpi.py:50]: Yes, this is how it's supposed to be. 15 is read, 16 is write, 17 is find offset.

File: revpi.py (hunk 3 above)
Comment (jarongrigat) [reply revpi.py]: ok

File: revpi.py (hunk 5 above)
Comment (jarongrigat) [reply revpi.py]: Done

File: revpi.py (hunk 4 above)
Comment (jarongrigat) [reply revpi.py]: Done

**Review by darrylmasson — APPROVED** [review]

**Pull request merged**